src/library_fs_shared.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * @license
+ * Copyright 2032 The Emscripten Authors
+ * SPDX-License-Identifier: MIT
+ */
+
+mergeInto(LibraryManager.library, {
+  // Preloads a file asynchronously. You can call this before run, for example in
+  // preRun. run will be delayed until this file arrives and is set up.
+  // If you call it after run(), you may want to pause the main loop until it
+  // completes, if so, you can use the onload parameter to be notified when
+  // that happens.
+  // In addition to normally creating the file, we also asynchronously preload
+  // the browser-friendly versions of it: For an image, we preload an Image
+  // element and for an audio, and Audio. These are necessary for SDL_Image
+  // and _Mixer to find the files in preloadedImages/Audios.
+  // You can also call this with a typed array instead of a url. It will then
+  // do preloading for the Image/Audio part, as if the typed array were the
+  // result of an XHR that you did manually.
+  $FS_createPreloadedFile__deps: ['$asyncLoad'],
+  $FS_createPreloadedFile: function(parent, name, url, canRead, canWrite, onload, onerror, dontCreateFile, canOwn, preFinish) {
+#if WASMFS
+    // TODO: use WasmFS code to resolve and join the path here?
+    var fullname = name ? parent + '/' + name : parent;
+#else
+    // TODO we should allow people to just pass in a complete filename instead
+    // of parent and name being that we just join them anyways
+    var fullname = name ? PATH_FS.resolve(PATH.join2(parent, name)) : parent;
```

Comment:
Do you know why this difference exists?  Is it just historical maybe?  Did PATH_FS not exist under wasmfs at some point?

Reply:
I don't know what PATH_FS is, but it doesn't sound like something that should be enabled under WasmFS. What's the difference in behavior between these two lines?

Reply:
I've not studied in detail but PATH_FS.resolve does a lot more work than just concatenation: https://github.com/emscripten-core/emscripten/blob/9d036213d1f13f3a6d2b6daef8069a93b66e1eda/src/library_path.js#L87-L105

For now I'll leave it as is.

Reply:
It looks like maybe it's doing some kind of path normalization that WasmFS doesn't need because it has better path parsing and can handle relative paths natively.